File: core/src/main/java/com/terralite/core/registry/GameData.java
package com.terralite.core.registry;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

public final class GameData {
    private final Map<RegistryKey<?>, FrozenRegistry<?>> registries;

    GameData(Map<RegistryKey<?>, FrozenRegistry<?>> registries) {
        this.registries = Collections.unmodifiableMap(new LinkedHashMap<>(Objects.requireNonNull(registries, "registries")));
    }

    public <T> FrozenRegistry<T> registry(RegistryKey<T> key) {
        FrozenRegistry<?> registry = registries.get(key);
        if (registry == null) {
            throw new IllegalArgumentException("Missing registry: " + key.id());
        }
        return cast(key, registry);
    }

    public <T> T get(ResourceKey<T> key) {
        return registry(key.registry()).require(key);
    }

    @SuppressWarnings("unchecked")
    private static <T> FrozenRegistry<T> cast(RegistryKey<T> key, FrozenRegistry<?> registry) {
        if (!registry.key().type().equals(key.type())) {
            throw new IllegalStateException("Registry type mismatch for " + key.id());
        }
        return (FrozenRegistry<T>) registry;
    }
}

File: core/src/commonMain/kotlin/com/terralite/core/registry/GameData.kt
package com.terralite.core.registry

/**
 * Immutable snapshot of all game registries, produced by [RegistryManager.freeze].
 *
 * Obtain a specific registry with [registry]; retrieve an entry directly with [get].
 */
class GameData internal constructor(
    private val registries: Map<RegistryKey<*>, FrozenRegistry<*>>
) {
    fun <T> registry(key: RegistryKey<T>): FrozenRegistry<T> {
        val registry = registries[key]
            ?: throw IllegalArgumentException("Missing registry: ${key.id}")
        return cast(key, registry)
    }

    fun <T> get(key: ResourceKey<T>): T = registry(key.registry).require(key)

    @Suppress("UNCHECKED_CAST")
    private fun <T> cast(key: RegistryKey<T>, registry: FrozenRegistry<*>): FrozenRegistry<T> {
        return registry as FrozenRegistry<T>
    }
}
